File: src/inference/engine.ts
import { join } from "node:path";
import { tmpdir } from "node:os";
import { randomUUID } from "node:crypto";
import { readFileSync, rmSync } from "node:fs";
import type { Config } from "../config.ts";

/**
 * The one place `ccs` runs LLM inference. Every call is structured (schema-forced),
 * non-agentic, and hermetic: pipe a bounded `stdin` payload, force a JSON `schemaPath`,
 * get back the parsed object or null. Two interchangeable backends sit behind one interface
 * — the Codex CLI (rides ChatGPT/OpenAI auth, no marginal cost) and the Claude Code CLI
 * (`claude -p`). See ADR-0001, which anticipated this swap being a one-file change.
 */
export type EngineName = "codex" | "claude";

export interface InferenceEngine {
  readonly name: EngineName;
  /** Whether the backing CLI is resolvable right now. Cached after the first probe. */
  available(): boolean;
  /**
   * Run one structured inference. `prompt` is the instruction, `stdin` the data payload,
   * `schemaPath` a JSON-Schema file the response must satisfy. Returns the parsed object
   * (caller casts to its expected shape) or null on any failure/timeout.
   */
  runStructured(opts: {
    prompt: string;
    stdin: string;
    schemaPath: string;
    timeoutMs?: number;
  }): Promise<unknown | null>;
}

/** Whether an executable is resolvable on PATH. Bun.spawn ignores shell aliases, so this
 *  mirrors what the engine will actually find when it spawns. */
export function binaryExists(binary: string): boolean {
  try {
    return Bun.spawnSync(["which", binary], { stdout: "ignore", stderr: "ignore" }).exitCode === 0;
  } catch {
    return false;
  }
}

/** Probe a binary once and cache the result — avoids a `which` per call. */
function lazyAvailability(binary: string): () => boolean {
  let cached: boolean | null = null;
  return () => {
    if (cached === null) cached = binaryExists(binary);
    return cached;
  };
}

/**
 * Codex backend. Runs `codex exec` hermetically (ephemeral, read-only sandbox, user
 * config/rules ignored) with `--output-schema` forcing the response and
 * `--output-last-message` capturing it to a temp file.
 */
export function createCodexEngine(opts: {
  binary: string;
  model: string;
  reasoningEffort: string;
}): InferenceEngine {
  const available = lazyAvailability(opts.binary);
  return {
    name: "codex",
    available,
    async runStructured({ prompt, stdin, schemaPath, timeoutMs = 60_000 }): Promise<unknown | null> {
      const outPath = join(tmpdir(), `ccs-infer-${randomUUID()}.json`);
      const args = [
        "exec",
        "--ephemeral",
        "--skip-git-repo-check",
        "--sandbox",
        "read-only",
        "--ignore-rules",
        "--ignore-user-config",
        "-c",
        `model_reasoning_effort="${opts.reasoningEffort}"`,
        "--output-schema",
        schemaPath,
        "--output-last-message",
        outPath,
      ];
      if (opts.model) args.push("-m", opts.model);
      args.push(prompt);

      try {
        const proc = Bun.spawn([opts.binary, ...args], {
          stdin: new TextEncoder().encode(stdin),
          stdout: "ignore",
          stderr: "ignore",
        });
        const timer = setTimeout(() => proc.kill(), timeoutMs);
        const code = await proc.exited;
        clearTimeout(timer);
        if (code !== 0) return null;
        return JSON.parse(readFileSync(outPath, "utf8")) as unknown;
      } catch {
        return null;
      } finally {
        rmSync(outPath, { force: true });
      }
    },
  };
}

/**
 * Claude Code backend. Runs `claude -p` non-interactively with `--json-schema` forcing a
 * structured response and `--output-format json`; the schema is passed inline (Claude wants
 * the JSON, not a path). `--strict-mcp-config` keeps the call cheap and hermetic by loading
 * no MCP servers. The parsed object comes back on the result envelope's `structured_output`.
 */
export function createClaudeEngine(opts: {
  binary: string;
  model: string;
}): InferenceEngine {
  const available = lazyAvailability(opts.binary);
  return {
    name: "claude",
    available,
    async runStructured({ prompt, stdin, schemaPath, timeoutMs = 60_000 }): Promise<unknown | null> {
      let schemaInline: string;
      try {
        // Collapse to a single line so it rides cleanly as one argv entry.
        schemaInline = JSON.stringify(JSON.parse(readFileSync(schemaPath, "utf8")));
      } catch {
        return null;
      }
      const args = ["-p", "--strict-mcp-config", "--output-format", "json", "--json-schema", schemaInline];
      if (opts.model) args.push("--model", opts.model);
      args.push(prompt);

      try {
        const proc = Bun.spawn([opts.binary, ...args], {
          stdin: new TextEncoder().encode(stdin),
          stdout: "pipe",
          stderr: "ignore",
        });
        const timer = setTimeout(() => proc.kill(), timeoutMs);
        const [out, code] = await Promise.all([new Response(proc.stdout).text(), proc.exited]);
        clearTimeout(timer);
        if (code !== 0) return null;

        const envelope = JSON.parse(out) as { structured_output?: unknown; result?: unknown };
        // Prefer the pre-parsed structured field; fall back to parsing the result string.
        if (envelope.structured_output && typeof envelope.structured_output === "object") {
          return envelope.structured_output;
        }
        return typeof envelope.result === "string" ? (JSON.parse(envelope.result) as unknown) : null;
      } catch {
        return null;
      }
    },
  };
}

/** Construct the engine for a given name from config. */
export function buildEngine(name: EngineName, config: Config): InferenceEngine {
  if (name === "claude") {
    return createClaudeEngine({ binary: config.inference.claude.binary, model: config.inference.claude.model });
  }
  return createCodexEngine({
    binary: config.inference.codex.binary,
    model: config.inference.codex.model,
    reasoningEffort: config.inference.codex.reasoningEffort,
  });
}

/** Preference order for `auto`: Codex first (rides free auth per ADR-0001), then Claude. */
const AUTO_ORDER: readonly EngineName[] = ["codex", "claude"];

/** Which engines are actually installed on this host, in preference order. */
export function detectAvailable(config: Config): EngineName[] {
  return AUTO_ORDER.filter((name) => buildEngine(name, config).available());
}

export interface EngineSelection {
  /** The engine to use now (best available). Null when nothing is installed. */
  readonly name: EngineName | null;
  /** Every installed engine, in preference order — drives whether the toggle is offered. */
  readonly available: EngineName[];
  /** What was requested (env > config), before availability fell it back. */
  readonly requested: EngineName | "auto";
  /** True when the requested engine wasn't installed and we fell back to another. */
  readonly fellBack: boolean;
}

const ENGINE_ENV = "CCS_INFERENCE_ENGINE";

function normalizeName(value: string | undefined | null): EngineName | "auto" | null {
  if (!value) return null;
  const v = value.trim().toLowerCase();
  return v === "codex" || v === "claude" || v === "auto" ? v : null;
}

/**
 * Resolve which engine to run with. Precedence: `CCS_INFERENCE_ENGINE` env → `override`
 * (e.g. a persisted TUI toggle) → `config.inference.engine` → `auto`. An explicit request
 * that isn't installed falls back to the best available one rather than failing.
 */
export function resolveEngine(config: Config, override?: EngineName | null): EngineSelection {
  const available = detectAvailable(config);
  const requested: EngineName | "auto" =
    normalizeName(process.env[ENGINE_ENV]) ?? override ?? normalizeName(config.inference.engine) ?? "auto";

  if (requested !== "auto" && available.includes(requested)) {
    return { name: requested, available, requested, fellBack: false };
  }
  const name = available[0] ?? null;
  return { name, available, requested, fellBack: requested !== "auto" && name !== null };
}

File: tui-go/inference/engine.go
// Package inference is the single structured model-call seam for ccs-go.
package inference

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"time"

	"github.com/pelletier/go-toml/v2"
)

// Name identifies one CCS-compatible inference backend.
type Name string

const (
	Codex  Name = "codex"
	Claude Name = "claude"
)

// Engine runs schema-forced, non-agentic inference through an installed CLI.
type Engine struct {
	Name            Name
	Binary          string
	Model           string
	ReasoningEffort string
}

type configFile struct {
	Inference struct {
		Engine string `toml:"engine"`
		Codex  struct {
			Binary          string `toml:"binary"`
			Model           string `toml:"model"`
			ReasoningEffort string `toml:"reasoningEffort"`
		} `toml:"codex"`
		Claude struct {
			Binary string `toml:"binary"`
			Model  string `toml:"model"`
		} `toml:"claude"`
	} `toml:"inference"`
}

// Resolve mirrors CCS engine.ts: env/config preference, otherwise Codex first and Claude fallback.
func Resolve() (Engine, error) {
	config, err := loadConfig()
	if err != nil {
		return Engine{}, err
	}
	codex := Engine{Name: Codex, Binary: "codex", ReasoningEffort: "low"}
	claude := Engine{Name: Claude, Binary: "claude", Model: "haiku"}
	if value := strings.TrimSpace(config.Inference.Codex.Binary); value != "" {
		codex.Binary = value
	}
	codex.Model = strings.TrimSpace(config.Inference.Codex.Model)
	if value := strings.TrimSpace(config.Inference.Codex.ReasoningEffort); value != "" {
		codex.ReasoningEffort = value
	}
	if value := strings.TrimSpace(config.Inference.Claude.Binary); value != "" {
		claude.Binary = value
	}
	if config.Inference.Claude.Model != "" {
		claude.Model = strings.TrimSpace(config.Inference.Claude.Model)
	}

	engines := []Engine{codex, claude}
	requested := normalizeName(os.Getenv("CCS_INFERENCE_ENGINE"))
	if requested == "" {
		requested = normalizeName(config.Inference.Engine)
	}
	if requested != "" && requested != "auto" {
		for _, engine := range engines {
			if string(engine.Name) == requested && engine.available() {
				return engine, nil
			}
		}
	}
	for _, engine := range engines {
		if engine.available() {
			return engine, nil
		}
	}
	return Engine{}, errors.New("no inference engine (codex/claude) found on PATH")
}

// RunStructured invokes one engine with an output schema and returns its decoded JSON object.
func (e Engine) RunStructured(ctx context.Context, prompt string, stdin string, schema []byte) (json.RawMessage, error) {
	if len(schema) == 0 {
		return nil, errors.New("empty output schema")
	}
	if !json.Valid(schema) {
		return nil, errors.New("invalid output schema")
	}
	switch e.Name {
	case Codex:
		return e.runCodex(ctx, prompt, stdin, schema)
	case Claude:
		return e.runClaude(ctx, prompt, stdin, schema)
	default:
		return nil, fmt.Errorf("unsupported inference engine %q", e.Name)
	}
}

func (e Engine) runCodex(ctx context.Context, prompt string, stdin string, schema []byte) (json.RawMessage, error) {
	dir, err := os.MkdirTemp("", "ccs-go-infer-")
	if err != nil {
		return nil, fmt.Errorf("create inference temp dir: %w", err)
	}
	defer os.RemoveAll(dir)
	schemaPath := filepath.Join(dir, "schema.json")
	outputPath := filepath.Join(dir, "output.json")
	if err := os.WriteFile(schemaPath, schema, 0o600); err != nil {
		return nil, fmt.Errorf("write inference schema: %w", err)
	}
	args := []string{
		"exec",
		"--ephemeral",
		"--skip-git-repo-check",
		"--sandbox", "read-only",
		"--ignore-rules",
		"--ignore-user-config",
		"-c", fmt.Sprintf("model_reasoning_effort=%q", e.ReasoningEffort),
		"--output-schema", schemaPath,
		"--output-last-message", outputPath,
	}
	if e.Model != "" {
		args = append(args, "-m", e.Model)
	}
	args = append(args, prompt)
	command := exec.CommandContext(ctx, e.Binary, args...)
	command.Stdin = strings.NewReader(stdin)
	command.Stdout = nil
	command.Stderr = nil
	if err := command.Run(); err != nil {
		return nil, commandError(e.Name, ctx, err)
	}
	output, err := os.ReadFile(outputPath)
	if err != nil {
		return nil, fmt.Errorf("read %s structured output: %w", e.Name, err)
	}
	if !json.Valid(output) {
		return nil, fmt.Errorf("%s returned invalid JSON", e.Name)
	}
	return json.RawMessage(output), nil
}

func (e Engine) runClaude(ctx context.Context, prompt string, stdin string, schema []byte) (json.RawMessage, error) {
	var compact bytes.Buffer
	if err := json.Compact(&compact, schema); err != nil {
		return nil, fmt.Errorf("compact inference schema: %w", err)
	}
	args := []string{
		"-p",
		"--no-session-persistence",
		"--strict-mcp-config",
		"--output-format", "json",
		"--json-schema", compact.String(),
	}
	if e.Model != "" {
		args = append(args, "--model", e.Model)
	}
	args = append(args, prompt)
	command := exec.CommandContext(ctx, e.Binary, args...)
	command.Stdin = strings.NewReader(stdin)
	var stdout bytes.Buffer
	command.Stdout = &stdout
	command.Stderr = nil
	if err := command.Run(); err != nil {
		return nil, commandError(e.Name, ctx, err)
	}
	var envelope struct {
		Structured json.RawMessage `json:"structured_output"`
		Result     json.RawMessage `json:"result"`
	}
	if err := json.Unmarshal(stdout.Bytes(), &envelope); err != nil {
		return nil, fmt.Errorf("decode claude result envelope: %w", err)
	}
	if len(envelope.Structured) > 0 && string(envelope.Structured) != "null" {
		return envelope.Structured, nil
	}
	var result string
	if json.Unmarshal(envelope.Result, &result) == nil && json.Valid([]byte(result)) {
		return json.RawMessage(result), nil
	}
	return nil, errors.New("claude returned no structured output")
}

func (e Engine) available() bool {
	if strings.TrimSpace(e.Binary) == "" {
		return false
	}
	_, err := exec.LookPath(e.Binary)
	return err == nil
}

func loadConfig() (configFile, error) {
	var config configFile
	root := strings.TrimSpace(os.Getenv("CCS_ROOT"))
	if root == "" {
		home, err := os.UserHomeDir()
		if err != nil {
			return config, fmt.Errorf("resolve home directory: %w", err)
		}
		root = filepath.Join(home, ".ccs")
	}
	contents, err := os.ReadFile(filepath.Join(root, "config.toml"))
	if errors.Is(err, os.ErrNotExist) {
		return config, nil
	}
	if err != nil {
		return config, fmt.Errorf("read inference config: %w", err)
	}
	if err := toml.Unmarshal(contents, &config); err != nil {
		return config, fmt.Errorf("parse inference config: %w", err)
	}
	return config, nil
}

func normalizeName(value string) string {
	switch strings.ToLower(strings.TrimSpace(value)) {
	case "codex":
		return "codex"
	case "claude":
		return "claude"
	case "auto":
		return "auto"
	default:
		return ""
	}
}

func commandError(name Name, ctx context.Context, err error) error {
	if errors.Is(ctx.Err(), context.DeadlineExceeded) {
		return fmt.Errorf("%s timed out", name)
	}
	return fmt.Errorf("%s inference failed: %w", name, err)
}

// WithTimeout creates the standard bounded inference context.
func WithTimeout(parent context.Context, duration time.Duration) (context.Context, context.CancelFunc) {
	if duration <= 0 {
		duration = 90 * time.Second
	}
	return context.WithTimeout(parent, duration)
}
